src/librustc_typeck/check/cast.rs

```diff
@@ -139,6 +140,21 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
 
     fn report_cast_error(&self, fcx: &FnCtxt<'a, 'gcx, 'tcx>, e: CastError) {
         match e {
+            CastError::NeedDeref => {
+                let cast_ty = fcx.ty_to_string(self.cast_ty);
+                let mut err = fcx.type_error_struct(self.cast_span,
+                                       |actual| {
+                                           format!("casting `{}` as `{}` is invalid",
+                                                   actual,
+                                                   cast_ty)
+                                       },
+                                       self.expr_ty);
+                err.span_label(self.cast_span,
+                               &format!("casting `{}` as `{}` is invalid", self.expr_ty, cast_ty));
+                err.span_label(self.expr.span,
```

Comment:
If we show the deref, it should be `span_help`.

Another way would be to change the label to, for example, "did you mean *{}?"